pydicom/pixel_data_handlers/util.py

```diff
@@ -220,23 +229,164 @@ def apply_modality_lut(arr, ds):
+        An array with applied VOI LUT or windowing operation.
+
+    References
+    ----------
+    * DICOM Standard, Part 3, :dcm:`Annex C.11.2
+      <part03/sect_C.11.html#sect_C.11.2>`
+    * DICOM Standard, Part 3, :dcm:`Annex C.8.11.3.1.5
+      <part03/sect_C.8.11.3.html#sect_C.8.11.3.1.5>`
+    * DICOM Standard, Part 4, :dcm:`Annex N.2.1.1
+      <part04/sect_N.2.html#sect_N.2.1.1>`
+    """
+    if 'VOILUTSequence' in ds:
+        # VOI LUT Sequence contains one or more items
+        item = ds.VOILUTSequence[index]
+        nr_entries = item.LUTDescriptor[0] or 2**16
+        first_map = item.LUTDescriptor[1]
+
+        # PS3.3 C.8.11.3.1.5: may be 8, 10-16
+        nominal_depth = item.LUTDescriptor[2]
+        if nominal_depth in list(range(10, 17)):
+            dtype = 'uint16'
+        elif nominal_depth == 8:
+            dtype = 'uint8'
+        else:
+            raise NotImplementedError(
+                "'{}' bits per LUT entry is not supported"
+                .format(nominal_depth)
+            )
+
+        lut_data = np.asarray(item.LUTData, dtype=dtype)
+
+        # IVs < `first_map` get set to first LUT entry (i.e. index 0)
+        clipped_iv = np.zeros(arr.shape, dtype=arr.dtype)
+        # IVs >= `first_map` are mapped by the VOI LUT
+        # `first_map` may be negative, positive or 0
+        mapped_pixels = arr >= first_map
+        clipped_iv[mapped_pixels] = arr[mapped_pixels] - first_map
+        # IVs > number of entries get set to last entry
+        np.clip(clipped_iv, 0, nr_entries - 1, out=clipped_iv)
+
+        return lut_data[clipped_iv]
+    elif 'WindowCenter' in ds and 'WindowWidth' in ds:
+        if ds.PhotometricInterpretation not in ['MONOCHROME1', 'MONOCHROME2']:
+            raise ValueError(
+                "When performing a windowing operation only 'MONOCHROME1' and "
+                "'MONOCHROME2' are allowed for (0028,0004) Photometric "
+                "Interpretation"
+            )
+
+        # May be LINEAR (default), LINEAR_EXACT, SIGMOID or not present, VM 1
+        voi_func = getattr(ds, 'VOILUTFunction', 'LINEAR').upper()
+        # VR DS, VM 1-n
+        elem = ds['WindowCenter']
+        center = elem.value[index] if elem.VM > 1 else elem.value
+        elem = ds['WindowWidth']
+        width = elem.value[index] if elem.VM > 1 else elem.value
+
+        # If output range depends on whether or not a rescale operation has
+        #   been applied - if not then the range is the available bit depth
+        if ds.PixelRepresentation:
+            y_min = -2**(ds.BitsStored - 1)
+            y_max = 2**(ds.BitsStored - 1) - 1
+        else:
+            y_min = 0
+            y_max = 2**ds.BitsStored - 1
+
+        if 'RescaleSlope' in ds and 'RescaleIntercept' in ds:
+            # Otherwise its the actual data range - see PS3.3 C.11.1.1.1
+            y_min = y_min * ds.RescaleSlope + ds.RescaleIntercept
+            y_max = y_max * ds.RescaleSlope + ds.RescaleIntercept
```

Comment:
Yes, its the output range of the modality LUT's rescale operation. That range is used to determine ymin and ymax for (PS3.3 C.11.1.1.1):

>These Attributes are applied according to the following pseudo-code, where x is the input value, y is an output value with a range from ymin to ymax

Essentially the windowing operation renormalises a subrange within the values so that it covers the full output range.

Reply:
Yes, you are right - I got this wrong. I'm still not completely sure if we should adapt `y_max` in case a modality LUT sequence has been applied, according to:
> In the case where the Modality LUT Sequence is used, the output range is from 0 to 2<sup>n</sup>-1 where n is the third value of LUT Descriptor.

Reply:
Argh, yes. I need to change it to account for that.

Reply:
Done